### PR TITLE
tm revamp: remove most topo.ChangeType

### PR DESCRIPTION
**Pull request by sougou**
All those calls have been replaced with agent.ChangeType.
agent.ChangeType is the only one that should call topo.ChangeType.
One exception is agent.finalizeTabletExternallyReparented. It
actually changes the tablet's internal state before updating
topo, which seems to break the main rule. So, I've left it alone
for now.

Signed-off-by: Sugu Sougoumarane <ssougou@gmail.com>

### CODE REVIEW

**Comment by deepthi** [comment]
We deprecated the TER vttablet RPC in 4.0 but forgot to remove it in 5.0/6.0. We can delete that and then finalize can also be deleted.

File: go/vt/vttablet/tabletmanager/rpc_external_reparent.go [hunk 1]
```diff
@@ -153,12 +153,15 @@ func (agent *ActionAgent) finalizeTabletExternallyReparented(ctx context.Context
 		defer wg.Done()
 		log.Infof("finalizeTabletExternallyReparented: updating tablet record for new master: %v", agent.TabletAlias)
 		// Update our own record to master if needed
+		// The internal state of agent has already been updated. No need to call RefreshState for this change.
```
[the remark this answers is not in the document]
Comment (deepthi) [reply go/vt/vttablet/tabletmanager/rpc_external_reparent.go:156]: On second thought, we should do delete/deprecation as a separate PR.

File: go/vt/vttablet/tabletmanager/rpc_backup.go [hunk 2]
```diff
@@ -84,12 +83,7 @@ func (agent *ActionAgent) Backup(ctx context.Context, concurrency int, logger lo
 		}
 		originalType = tablet.Type
 		// update our type to BACKUP
-		if _, err := topotools.ChangeType(ctx, agent.TopoServer, tablet.Alias, topodatapb.TabletType_BACKUP, nil); err != nil {
-			return err
-		}
-
-		// let's update our internal state (stop query service and other things)
-		if err := agent.refreshTablet(ctx, "before backup"); err != nil {
+		if err := agent.ChangeType(ctx, topodatapb.TabletType_BACKUP); err != nil {
```
Comment (deepthi) [inline go/vt/vttablet/tabletmanager/rpc_backup.go:86]: This is a problem. We are calling an RPC within an RPC, so it's blocking on the actionMutex.

File: go/vt/vttablet/tabletmanager/rpc_backup.go [hunk 3]
```diff
@@ -120,8 +114,7 @@ func (agent *ActionAgent) Backup(ctx context.Context, concurrency int, logger lo
 
 		// Change our type back to the original value.
 		// Original type could be master so pass in a real value for masterTermStartTime
-		_, err = topotools.ChangeType(bgCtx, agent.TopoServer, tablet.Alias, originalType, tablet.Tablet.MasterTermStartTime)
-		if err != nil {
+		if err := agent.ChangeType(ctx, originalType); err != nil {
```
Comment (deepthi) [inline go/vt/vttablet/tabletmanager/rpc_backup.go:117]: The next call to refreshTablet is unnecessary, because ChangeType already calls it.

File: go/vt/vttablet/tabletmanager/rpc_backup.go (hunk 2 above)
Comment (deepthi) [reply go/vt/vttablet/tabletmanager/rpc_backup.go:86]: Another approach would be to create a `changeTypeLocked` which is called when we are already holding the action lock.

Comment (enisoc) [reply go/vt/vttablet/tabletmanager/rpc_backup.go:82]: I think we were taking the lock in this block so that, in the case of drain-for-backup, we exclude all other RPC actions while the backup is happening. If that's true, we still need to hold the lock throughout the rest of this function, not just while changing type. The pattern we typically use in Vitess for this is to pull the meat of ChangeType() out into changeTypeLocked() and call the latter whenever we already hold the lock.

File: go/vt/vttablet/tabletmanager/rpc_replication.go [hunk 4]
```diff
@@ -194,11 +192,6 @@ func (agent *ActionAgent) ResetReplication(ctx context.Context) error {
 
 // InitMaster enables writes and returns the replication position.
 func (agent *ActionAgent) InitMaster(ctx context.Context) (string, error) {
-	if err := agent.lock(ctx); err != nil {
```
Comment (enisoc) [inline go/vt/vttablet/tabletmanager/rpc_replication.go:197]: This looks unsafe. There's a lot happening now without holding the lock. The current goal of the lock is to serialize entire RPCs, not just certain changes. I don't think we can relax that without thinking through a full redesign of the locking scheme.

File: go/vt/vttablet/tabletmanager/rpc_replication.go [hunk 5]
```diff
@@ -260,10 +242,14 @@ func (agent *ActionAgent) PopulateReparentJournal(ctx context.Context, timeCreat
 // InitSlave sets replication master and position, and waits for the
 // reparent_journal table entry up to context timeout
 func (agent *ActionAgent) InitSlave(ctx context.Context, parent *topodatapb.TabletAlias, position string, timeCreatedNS int64) error {
-	if err := agent.lock(ctx); err != nil {
-		return err
+	// If we were a master type, switch our type to replica.  This
+	// is used on the old master when using InitShardMaster with
+	// -force, and the new master is different from the old master.
+	if agent.Tablet().Type == topodatapb.TabletType_MASTER {
```
Comment (enisoc) [inline go/vt/vttablet/tabletmanager/rpc_replication.go:248]: Doing this check without holding the lock seems unsafe. We won't know after we finally acquire the lock inside ChangeType whether the state we checked here still holds.

File: go/vt/vttablet/tabletmanager/action_agent.go [hunk 6]
```diff
@@ -159,10 +159,6 @@ type ActionAgent struct {
 	// instead of asking mysqld what port it's serving on.
 	mysqlAdvertisePort int32
 
-	// initReplication remembers whether an action has initialized
-	// replication.  It is protected by actionMutex.
-	initReplication bool
```
Comment (enisoc) [inline go/vt/vttablet/tabletmanager/action_agent.go:164]: Looks like this was being assigned but never read? Just making sure I understand why it's being removed.

File: go/vt/vttablet/tabletmanager/rpc_replication.go (hunk 4 above)
Comment (deepthi) [reply go/vt/vttablet/tabletmanager/rpc_replication.go:197]: +1

File: go/vt/vttablet/tabletmanager/rpc_backup.go (hunk 2 above)
Comment (deepthi) [reply go/vt/vttablet/tabletmanager/rpc_backup.go:82]: Good point. I forgot that defer only gets executed at the end of the function, not at the end of the block.

File: go/vt/vttablet/tabletmanager/action_agent.go (hunk 6 above)
Comment (sougou) [reply go/vt/vttablet/tabletmanager/action_agent.go:164]: Yes.

File: go/vt/wrangler/testlib/rpc_reparent_external_test.go [hunk 7]
```diff
@@ -292,6 +292,10 @@ func TestRPCTabletExternallyReparentedFailedOldMaster(t *testing.T) {
 		"START SLAVE",
 	}
 
+	// On the oldMaster, we expect a ChangeType to REPLICA.
```
Comment (deepthi) [inline go/vt/wrangler/testlib/rpc_reparent_external_test.go:295]: Why were these changes needed?

Comment (sougou) [reply go/vt/wrangler/testlib/rpc_reparent_external_test.go:295]: Because we're now using `agent.ChangeType` instead of directly updating the topo. So, the ActionAgent needs to be initialized for it to process `ChangeType` properly.

Comment (deepthi) [reply go/vt/wrangler/testlib/rpc_reparent_external_test.go:295]: You could revert all changes to `rpc_external_reparent.go` and this test and delete the functions in that file + tests either in this PR or in a subsequent one. It doesn't make sense to make changes to it when we should be deleting it.

Comment (sougou) [reply go/vt/wrangler/testlib/rpc_reparent_external_test.go:295]: It's fine. I'll ping you separately for the cleanup.

**Comment by sougou** [comment]
local_example has been failing, but it passes on my machine. I'll keep retrying.
